docs/src/main/paradox/grpc.md

```diff
@@ -135,6 +135,132 @@ Java
+consumer. The transformation function on the producer side can omit certain events, but the offsets for these
+events are still transferred to the consumer, to ensure sequence number validations and offset storage.
+
+Filters can be used when a consumer is only interested in a subset of the entities. The filters can be defined
+on both the producer side and on the consumer side, and they can be changed at runtime.
+
+### Tags
+
+Tags are typically used for the filters, so first an example of how to tag events in the entity. Here, the tag is
+based on total quantity of the shopping cart, i.e. the state of the cart. The tags are included in the
+@apidoc[akka.persistence.query.typed.EventEnvelope].
+
+Scala
+:  @@snip [ShoppingCart.scala](/samples/grpc/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala) { #tags }
+
+Java
+:  @@snip [ShoppingCart.java](/samples/grpc/shopping-cart-service-java/src/main/java/shopping/cart/ShoppingCart.java) { #tags }
+
+
+### Producer defined filter
+
+The producer may define a filter function on the `EventProducerSource`.
+
+Scala
+:  @@snip [PublishEvents.scala](/samples/grpc/shopping-cart-service-scala/src/main/scala/shopping/cart/PublishEvents.scala) { #withProducerFilter }
+
+Java
+:  @@snip [PublishEvents.java](/samples/grpc/shopping-cart-service-java/src/main/java/shopping/cart/PublishEvents.java) { #withProducerFilter }
+
+In this example the decision is based on tags, but the filter function can use anything in the
+@apidoc[akka.persistence.query.typed.EventEnvelope] parameter or the event itself. Here, the entity sets the tag based
+on the total quantity of the shopping cart, which requires the full state of the shopping cart and is not known from
+an individual event.
+
+Note that the purpose of the `withProducerFilter` is to toggle if all events for the entity is to be emitted or not.
+If the purpose is to filter out certain events you should instead use the `Transformation`.
+
+The producer filter is evaluated before the transformation function, i.e. the event is the original event and not
+the transformed event.
+
+A producer filter that excludes an event wins over any consumer defined filter, i.e. if the producer filter function
+returns `false` the event will not be emitted.
+
+### Consumer defined filter
+
+The consumer may define declarative filters that are sent to the producer and evaluated on the producer side
+before emitting the events.
+
+Consumer filters consists of exclude and include criteria. In short, the exclude criteria are evaluated first and
+may be overridden by an include criteria. More precisely, they are evaluated according to the following rules: 
+
+* Exclude criteria are evaluated first.
+* If no matching exclude the event is emitted.
+* If an exclude is matching the include criteria are evaluated.
+* If no matching include the event is discarded.
+* If matching include the event is emitted.
```

Comment:
Can be confusing with words "exclude" and "include" are used as nouns. 

We can append "criterion" /  "criteria" to it. 

```suggestion
* Exclude criteria are evaluated first.
* If no matching exclude criterion, then the event is emitted.
* If an exclude criterion matches, then the include criteria are evaluated.
* If no matching include criterion, then the event is discarded.
* If matching include criterion, then the event is emitted.
```

or

```suggestion
* Exclude criteria are evaluated first.
* If no matching `exclude`, then the event is emitted.
* If an `exclude` is matching, then the `include` criteria are evaluated.
* If no matching `include`, then the event is discarded.
* If matching `include`, then the event is emitted.
```

Reply:
sounds good to add criteria. I think I prefer criteria over criterion. Both are [grammatically correct](https://www.merriam-webster.com/words-at-play/criteria-vs-criterion-singular-plural-grammar). criteria can be used as both singular and plural

Reply:
added in https://github.com/akka/akka-projection/pull/846/commits/6da5e7862a9d1b238d038e8bce3d19ed6f1c7d75

Reply:
Looks good.